design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+# Contour Operator Gateway API Support Migration
+
+Status: Draft
+
+## Abstract
+Currently the Contour Operator and Contour Controller have both started implementing Gateway API support.
+Both components contain controllers that reconcile GatewayClass and Gateway resources and that conflict when reconciling status.
+This document describes options for how we should proceed to untangle these.
+
+## Background
+Currently the state of the world is as below:
+
+### Contour Operator
+- Configured with a GatewayClass object reference (name) to reconcile
+- Includes GatewayClass controller
+  - Checks if named GatewayClass has the correct (hardcoded to `projectcontour.io/contour-operator`) controller string
+  - Checks for Contour CRD in the parameters ref of the GatewayClass, uses parameters for provisioning Contour instances
+  - Sets admitted status on object if all valid
+- Includes Gateway controller
+  - Checks if Gateway refers to correct GatewayClass and that GatewayClass has been admitted
+  - Sets up finalizer for Gateway
+  - Creates Contour deployment resources with the Gateway being reconciled named in Contour's ConfigMap
```

Comment:
When exactly does Contour get created?  Is the user doing this by hand? We always the Gateway is a contract for provisioning the infrastructure so does the existence of a Gateway immediately trigger the Operator to provision a kind:Contour (set of Contours + Envoys)

Reply:
> Is the user doing this by hand? We always the Gateway is a contract for provisioning the infrastructure so does the existence of a Gateway immediately trigger the Operator to provision a kind:Contour (set of Contours + Envoys)

A user creates a Contour CRD instance by hand and the Operator Gateway controller uses that + the existence of a Gateway (also created by a user by hand) to create a Contour Deployment+Envoy DaemonSet